builtin/providers/dockercloud/resource_dockercloud_node_cluster.go

```diff
@@ -0,0 +1,255 @@
+
+		Schema: map[string]*schema.Schema{
+			"name": &schema.Schema{
+				Type:     schema.TypeString,
+				Required: true,
+				ForceNew: true,
+			},
+			"node_provider": &schema.Schema{
+				Type:     schema.TypeString,
+				Required: true,
+				ForceNew: true,
+			},
+			"size": &schema.Schema{
+				Type:     schema.TypeString,
+				Required: true,
+				ForceNew: true,
+			},
+			"region": &schema.Schema{
+				Type:     schema.TypeString,
+				Required: true,
+				ForceNew: true,
+			},
+			"disk": &schema.Schema{
+				Type:     schema.TypeInt,
+				Optional: true,
+				Computed: true,
+				ForceNew: true,
+			},
+			"node_count": &schema.Schema{
+				Type:     schema.TypeInt,
+				Optional: true,
+				Computed: true,
+			},
+			"tags": &schema.Schema{
+				Type:     schema.TypeList,
+				Optional: true,
+				Elem:     &schema.Schema{Type: schema.TypeString},
+			},
+		},
+	}
+}
+
+func resourceDockercloudNodeClusterCreate(d *schema.ResourceData, meta interface{}) error {
+	provider := d.Get("node_provider").(string)
+	region := d.Get("region").(string)
+	size := d.Get("size").(string)
+
+	opts := dockercloud.NodeCreateRequest{
+		Name:     d.Get("name").(string),
+		Region:   fmt.Sprintf("/api/infra/v1/region/%s/%s/", provider, region),
+		NodeType: fmt.Sprintf("/api/infra/v1/nodetype/%s/%s/", provider, size),
+	}
+
+	if attr, ok := d.GetOk("disk"); ok {
+		opts.Disk = attr.(int)
+	}
+
+	if attr, ok := d.GetOk("node_count"); ok {
+		opts.Target_num_nodes = attr.(int)
+	}
+
+	tags := d.Get("tags.#").(int)
+	if tags > 0 {
+		opts.Tags = make([]dockercloud.NodeTag, 0, tags)
+		for i := 0; i < tags; i++ {
+			key := fmt.Sprintf("tags.%d", i)
+			opts.Tags = append(opts.Tags, dockercloud.NodeTag{Name: d.Get(key).(string)})
+		}
+	}
+
+	nodeCluster, err := dockercloud.CreateNodeCluster(opts)
+	if err != nil {
+		return err
+	}
+
+	if err = nodeCluster.Deploy(); err != nil {
+		return fmt.Errorf("Error creating node cluster: %s", err)
+	}
+
+	d.SetId(nodeCluster.Uuid)
+
+	stateConf := &resource.StateChangeConf{
+		Pending:        []string{"Deploying"},
+		Target:         []string{"Deployed"},
+		Refresh:        newNodeClusterStateRefreshFunc(d, meta),
+		Timeout:        60 * time.Minute,
+		Delay:          10 * time.Second,
+		MinTimeout:     3 * time.Second,
+		NotFoundChecks: 60,
+	}
+
+	_, err = stateConf.WaitForState()
+	if err != nil {
+		return fmt.Errorf("Error waiting for node cluster (%s) to become ready: %s", d.Id(), err)
+	}
+
+	return resourceDockercloudNodeClusterRead(d, meta)
+}
+
+func resourceDockercloudNodeClusterRead(d *schema.ResourceData, meta interface{}) error {
+	nodeCluster, err := dockercloud.GetNodeCluster(d.Id())
+	if err != nil {
+		if strings.Contains(err.Error(), "404 NOT FOUND") {
+			d.SetId("")
+			return nil
+		}
+
+		return fmt.Errorf("Error retrieving node cluster: %s", err)
+	}
+
+	if nodeCluster.State == "Terminated" {
+		d.SetId("")
+		return nil
+	}
+
+	d.Set("name", nodeCluster.Name)
```

Comment:
They use URIs for region and size/type. I added `node_provider` as a sort of convenience to build the URI for region and size/type rather than have specify it explicitly (they hide these away behind the API). Setting the state requires a bit of ugly parsing of the URIs returned. I will add a function to parse these URIs, but am on the fence of just abandoning building URIs altogether and simply just documenting the current URI structure and having them passed in as `/api/infra/v1/region/aws/us-east-1/` for example.